src/ucp/core/ucp_ep.c

```diff
@@ -233,58 +233,167 @@ ucp_ep_adjust_params(ucp_ep_h ep, const ucp_ep_params_t *params)
 ucs_status_t ucp_worker_create_mem_type_endpoints(ucp_worker_h worker)
 {
     ucp_context_h context = worker->context;
-    ucp_unpacked_address_t local_address;
+    uint64_t tl_bitmap;
     unsigned i, mem_type;
     ucs_status_t status;
-    void *address_buffer;
-    size_t address_length;
 
     for (mem_type = 0; mem_type < UCS_MEMORY_TYPE_LAST; mem_type++) {
-        if (UCP_MEM_IS_ACCESSIBLE_FROM_CPU(mem_type) ||
-            !context->mem_type_access_tls[mem_type]) {
+        tl_bitmap = context->mem_type_access_tls[mem_type];
+        if (UCP_MEM_IS_ACCESSIBLE_FROM_CPU(mem_type) || !tl_bitmap) {
             continue;
         }
 
-        status = ucp_address_pack(worker, NULL,
-                                  context->mem_type_access_tls[mem_type],
-                                  UCP_ADDRESS_PACK_FLAG_ALL, NULL,
-                                  &address_length, &address_buffer);
+        /* EP to access accelerators memory like CUDA/ROCM/etc */
+        status = ucp_worker_create_loopback_ep(worker, tl_bitmap,
+                                               UCP_EP_INIT_FLAG_MEM_TYPE,
+                                               "mem type",
+                                               &worker->mem_type_ep[mem_type]);
         if (status != UCS_OK) {
             goto err_cleanup_eps;
         }
+    }
 
-        status = ucp_address_unpack(worker, address_buffer, UINT64_MAX, &local_address);
-        if (status != UCS_OK) {
-            goto err_free_address_buffer;
+    return UCS_OK;
+
+err_cleanup_eps:
+    for (i = 0; i < UCS_MEMORY_TYPE_LAST; i++) {
+        if (worker->mem_type_ep[i]) {
```

Comment:
is it possible that mem_type_ep is uninitialized?

Reply:
no, worker object is allocated by `ucs_calloc`